File: rendering/textures/ProceduralTextureGenerator.cs
using System;

namespace PiroEngine
{
    public class ProceduralTextureGenerator
    {
        public Texture GeneratePerlinNoiseTexture(int width, int height, float scale)
        {
            Texture texture = new Texture(width, height);

            for (int y = 0; y < height; y++)
            {
                for (int x = 0; x < width; x++)
                {
                    float xCoord = (float)x / width * scale;
                    float yCoord = (float)y / height * scale;
                    
                    float noiseValue = PerlinNoise(xCoord, yCoord);
                    
                    Color color = new Color(noiseValue, noiseValue, noiseValue);
                    texture.SetPixel(x, y, color);
                }
            }

            return texture;
        }

        private float PerlinNoise(float x, float y)
        {
            // Реализация алгоритма Perlin Noise здесь
            // ...
        }
    }
}

File: textures/procedural_texture_generator.py
import numpy as np

class ProceduralTextureGenerator:
    def __init__(self, width, height):
        self.width = width
        self.height = height

    def generate_perlin_noise_texture(self, scale=10, octaves=6, persistence=0.5):
        noise_texture = np.zeros((self.width, self.height))
        for octave in range(octaves):
            octave_scale = scale * (2 ** octave)
            octave_noise = self._generate_perlin_noise(octave_scale)
            noise_texture += octave_noise * persistence
            persistence *= 0.5  # Decrease persistence with each octave
        return noise_texture

    def _generate_perlin_noise(self, scale):
        # Generate Perlin noise for the given scale
        # (Implementation of Perlin noise generation goes here)
        pass

    def generate_marble_texture(self, scale=5):
        # Generate a marble texture using Perlin noise
        marble_texture = self.generate_perlin_noise_texture(scale, octaves=8, persistence=0.6)
        # Apply a sine wave to create marble-like patterns
        marble_texture = np.sin(marble_texture)
        return marble_texture
